classes/services/PKPFileService.inc.php

```diff
@@ -117,22 +117,35 @@ public function delete($id) {
 	 * This method sends a HTTP response and ends the request handling.
 	 * No code will run after this method is called.
 	 *
-	 * @param string $path The path to the file
+	 * @param string|int $pathOrFileId The path to the file or file ID
 	 * @param string $filename Filename to give to the downloaded file
 	 * @param boolean $inline Whether to stream the file to the browser
 	 */
-	public function download($path, $filename, $inline = false) {
+	public function download($pathOrFileId, $filename, $inline = false) {
+
+		$dispatcher = Application::get()->getRequest()->getDispatcher();
+
+		if (is_integer($pathOrFileId)) {
+			// Is the file ID
+			$fileId = $pathOrFileId;
+			$file = $this->get($fileId);
+			if (!$file) $dispatcher->handle404();
+			$path = $file->path;
+		} else {
+			// Is the path to the file; compatibility fix, see pkp/pkp-lib#6663
+			$path = $pathOrFileId;
+		}
 
 		if (!$this->fs->has($path)) {
-			Application::get()->getRequest()->getDispatcher()->handle404();
+			$dispatcher->handle404();
 		}
 
-		if (HookRegistry::call('File::download', [$path, &$filename, $inline])) {
+		if (HookRegistry::call('File::download', [$file ?? $path, &$filename, $inline])) {
 			return;
 		}
 
 		// Stream the file to the end user.
-		$mimetype = $this->fs->getMimetype($path) ?? 'application/octet-stream';
+		$mimetype = $file->mimetype ?? $this->fs->getMimetype($path) ?? 'application/octet-stream';
```

Comment:
Did you test this with a path? I think calling `$file->mimetype` willl cause an error if only the path was called...

Reply:
Yes, tested it. It first will check if `$file` and it's mimetype property is initialized, before null check; shouldn't cause a problem.

Reply:
Thanks, you're right. I didn't believe it but the null coalescing operator works on object properties even when the object isn't defined. :tada: